build.gradle.kts

```diff
@@ -103,6 +110,54 @@ val requireNativeTask by tasks.register("requireNativeDirectory") {
 	}
 }
 
-tasks.withType<KotlinCompile>().all {
+val downloadJExtractTask by tasks.register<Download>("downloadJExtract") {
+	val os = DefaultNativePlatform.getCurrentOperatingSystem()
+	val url = when {
+		os.isMacOsX -> "https://download.java.net/java/early_access/jextract/1/openjdk-21-jextract+1-2_macos-x64_bin.tar.gz"
+		os.isLinux -> "https://download.java.net/java/early_access/jextract/1/openjdk-21-jextract+1-2_linux-x64_bin.tar.gz"
+		else -> throw IllegalStateException("Unsupported OS: ${os.name}.")
+	}
+
+	src(url)
+	onlyIfModified(true)
+	useETag(true)
+	dest(layout.buildDirectory.file("resources/jextract.tar.gz").get().asFile)
+}
+
+val extractJExtractTask by tasks.register<Copy>("extractJExtract") {
```

Comment:
All of these `by`s should be `=` as you're forcing eager creation of the tasks right now.

Inside the configuration block of another task you can either `.map { .. }` it to something you want if you can accept a provider, or just call `.get()` to get the task at that point.

Reply:
Ah ha! This is the key thing I was missing for implicit task dependencies. Using `=` and `map { }` lets me remove the `dependsOn`s.